File: rekall-model/src/main/java/dev/rekall/domain/CommitReference.java
package dev.rekall.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Lob;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

@Entity
@Table(name = "commit_reference")
@Getter
public class CommitReference {

    public static final int HASH_MAX = 40;
    public static final int COMMENT_MAX = 200;
    public static final int DIFF_MAX = 200_000;

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(
            name = "task_id",
            nullable = false,
            updatable = false,
            foreignKey = @ForeignKey(name = "fk_commit_reference_task"))
    private Task task;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "step_id",
            updatable = false,
            foreignKey = @ForeignKey(name = "fk_commit_reference_step"))
    private TaskStep step;

    @Column(name = "commit_hash", nullable = false, updatable = false, length = HASH_MAX)
    private String commitHash;

    @Column(name = "comment", nullable = false, updatable = false, length = COMMENT_MAX)
    private String comment;

    /** The diff this commit introduced, so the change can be reread later without a checkout. Null for a commit logged before this column existed, or when git could not produce one. */
    @Lob
    @Column(name = "diff", updatable = false)
    private String diff;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    protected CommitReference() {
    }

    public CommitReference(Task task, TaskStep step, String commitHash, String comment, String diff) {
        this.task = task;
        this.step = step;
        this.commitHash = commitHash;
        this.comment = comment;
        this.diff = diff;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof CommitReference that && id != null && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "CommitReference[" + commitHash + " " + task.getLabel() + "]";
    }
}

File: rekall-model/src/commit_reference.rs
//! A commit of the project's folder logged against a task, or one of its steps, with the diff it
//! introduced so the change can be reread without a checkout.

use rekall_common::{Id, Instant, RekallError};
use sea_orm::entity::prelude::*;

use crate::constraints::{Phase, Violations};

pub const HASH_MAX: usize = 40;
pub const COMMENT_MAX: usize = 200;
pub const DIFF_MAX: usize = 200_000;

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "commit_reference")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Id,
    pub task_id: Id,
    pub step_id: Option<Id>,
    pub commit_hash: String,
    pub comment: String,
    /// Null for a commit logged before the column existed, or when git could not produce one.
    #[sea_orm(column_type = "Text", nullable)]
    pub diff: Option<String>,
    /// Whether this commit rides along with `rekall_context`. Off when logged; the console flips
    /// it, one row at a time.
    pub in_context: bool,
    pub created_at: Instant,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::task::Entity",
        from = "Column::TaskId",
        to = "super::task::Column::Id",
        on_delete = "Cascade"
    )]
    Task,
    #[sea_orm(
        belongs_to = "super::task_step::Entity",
        from = "Column::StepId",
        to = "super::task_step::Column::Id",
        on_delete = "Cascade"
    )]
    TaskStep,
}

impl Related<super::task::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Task.def()
    }
}

impl Related<super::task_step::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::TaskStep.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

impl Model {
    pub fn validate(&self) -> Result<(), RekallError> {
        Violations::new("CommitReference", Phase::Persist)
            .column("commit_hash", Some(&self.commit_hash), HASH_MAX)
            .column("comment", Some(&self.comment), COMMENT_MAX)
            .finish()
    }
}
